### PR TITLE
Mving cpp into rust

### PR DESCRIPTION
**Pull request by day01**
**Description**
This PR migrates the implementation from C++ to Rust to achieve WebAssembly (WASM) compatibility. This is a crucial step towards modernizing bs codebase, enhancing performance, and improving security.

@hayden4r4 
Feel free to add any additional details specific to your project. If you have any other suggestions or need further adjustments, let me know!

I am leaving open PR to discuss about it.

### CODE REVIEW

**Comment by hayden4r4** [comment]
awesome that you are working on this, this is something I've been wanting to do also, the most important part for me is benchmarking the performance between the rust and cpp implementation. Jackel's cpp implementation can solve for IV in 2 iterations for all inputs, we should ensure that we can preserve that along with maintaining similar (if not better) performance

**Comment by day01** [comment]
Currently i got the same results in 2 iteration. I have problem with fmt bcs so_rational is so_complicated :D 
About timing and possibility of optimalization we can improve it after first mving to rust and step by step.

Jackel's solution simple. So at the first step I recommend:
- moving solution to rust with 2 iteration
- optimalization of mem allocation n time execution
- introduce what @claireratigan starts in #5 but for jackel. I think it will be great option to introduce calculation on U256 :) 
- n of course moving solution to f64, everyone can cast f64 to f32 with cutting precision.

File: src/lets_be_rational/black.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+#[rustfmt::skip]
+pub fn asymptotic_expansion_of_normalised_black_call(h: f64, t: f64) -> f64 {
+    let e = (t / h) * (t / h);
+    let r = (h + t) * (h - t);
+    let q = (h / r) * (h / r);
+
+    // 17th order asymptotic expansion of A(h,t) in q
+    let asymptotic_expansion_sum = 2.0 + q * (-6.0 - 2.0 * e + 3.0 * q * (10.0 + e * (20.0 + 2.0 * e)
```
Comment (day01) [inline src/lets_be_rational/black.rs:111]: I don't think so it is good idea  to use predefined values. @hayden4r4 what do you think about Newton-Raphson method to use here?

**Comment by day01** [comment]
Solution has lot of unused method migrated from cpp, i am not sure which methods should be exposed officially which are unused and which one are to drop :D 

I think it is important part to doc part and i will need a little help with docs here. 
Personally I prefer minimal public api, so in my opinion it is enough to expose as is. It will be good if sth can confirm it

File: src/lets_be_rational/intrinsic.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        assert_relative_eq!(result, EXPECTED_RESULT, epsilon = EPSILON);
+    }
+
+    #[test]
+    fn test_normalised_intrinsic_with_large_x() {
+        let x = f64::MAX;
+        let option_type = OptionType::Call;
+        let result = normalised_intrinsic(x, option_type);
+
+        // TODO: to analyze correct paths here
```
Comment (day01) [inline src/lets_be_rational/intrinsic.rs:58]: for f64:MAX we have inf+. in my opinion we shouldn't get inf. I think we should consider some of error of different behaviour on another level. Or we should return Option

File: src/lets_be_rational/rational_cubic.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    d_l: f64,
+    d_r: f64,
+    r: f64,
+) -> f64 {
+    let h = x_r - x_l;
+    if h.abs() <= 0.0 {
+        return 0.5 * (y_l + y_r);
+    }
+    // r should be greater than -1.
+    // We do not use assert(r > -1) here in order to allow values such as NaN to be propagated as they should.
```
Comment (day01) [inline src/lets_be_rational/rational_cubic.rs:21]: in my opinion these implementation of cubic is terrible...

**Comment by day01** [comment]
@hayden4r4 imo rust solution is good enough.
Of course next steps is more test, more error handling and increase of performance n perf tests.

Please make a review and if it will get approve i will take next steps in solution.

File: src/lets_be_rational/black.rs (hunk 1 above)
Comment (hayden4r4) [reply src/lets_be_rational/black.rs:111]: I was using newton raphson originally but moved to this for reasons I can't remember, I am fine with moving it back to newton raphson to give an easier to understand/debug alternative to lets be rational

Comment (hayden4r4) [reply src/lets_be_rational/black.rs:111]: while this makes for ugly code I'm a fan of what gives us better performance here, I'm curious what the added runtime cost of using newton raphson here would be

**Comment by hayden4r4** [comment]
> Solution has lot of unused method migrated from cpp, i am not sure which methods should be exposed officially which are unused and which one are to drop :D
> 
> I think it is important part to doc part and i will need a little help with docs here. Personally I prefer minimal public api, so in my opinion it is enough to expose as is. It will be good if someone can confirm it

Given that you are doing the work of porting jackels code, it may be worth it to publish it as a seperate repo that we can bring in as a submodule.  Others may be able to make use of the methods in it that we don't need to expose publicly for this project.  I'd prefer keeping all the methods in it private and just exposing access to it through calc_rational_x methods like we have currently.

**Comment by hayden4r4** [comment]
I just pushed a new commit with rational IV benchmarking and ran benches on the cpp vs the rust implementation. Interestingly the runtime went from 254.16ns to 69.1ns with the rust port, a 72.6% improvement.  Really good stuff.

File: src/lets_be_rational/rational_cubic.rs (hunk 3 above)
Comment (hayden4r4) [reply src/lets_be_rational/rational_cubic.rs:21]: I think we should merge as is and look to replace this later if improvements can be made

**Review by hayden4r4 — APPROVED** [review]

**Comment by hayden4r4** [comment]
I merged this a little too quickly, reverted it now, a new PR will need to be opened as I can't reopen this after being merged.  It failed 4/5 of the new IV tests, with only the ATM call test passing.  The put ATM/OTM are somewhat close but outside of tolerance.  The ITM put/call are both > 1 point off so seems to be an issue there.

**Comment by day01** [comment]
@hayden4r4 thx for tests for iv. bug is fixed in #21. I made a optimalisation `misspelling` sign was wrong in last fixes xD

**Pull request merged**